### ISSUE TITLE
[HOLD] DEV - `OpenPaymentsPage` is called many times when opening a workspace chat

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#38721]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** Only in Dev
**Reproducible in staging?:** n/a
**Reproducible in production?:** n/a
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @youssef-lr 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1710976232831539

## Action Performed:
1. Open app
2. Open any workspace chat with IOU
## Expected Result:
There should be one api call for `OpenPaymentsPage`
## Actual Result:
`OpenPaymentsPage` called many times
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/01b842ec-622e-43b2-88f7-2e5b80ac40db

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- DEV - OpenPaymentsPage is called many times when opening a workspace chat
### What is the root cause of that problem?
- We have a logic to call ```openWalletPage``` foreach ```SettlementButton```:
https://github.com/Expensify/App/blob/47bba81007ae9a4a5dfd6008c626efecb895153d/src/components/SettlementButton.tsx#L134-L136
### What changes do you think we should make in order to solve the problem?
- We should move this logic to the report that contains these ``````SettlementButton```. To do that:
- First, create a function, add the below logic to ReportScreen:
```
    const didCallOpenWalletPage = useRef(false)
    useEffect(() => {
        if(didCallOpenWalletPage.current){
            return
        }
        const shouldCallOpenWalletPage = some(Object.values(reportActions), (action) => {
            if (action.actionName !== CONST.REPORT.ACTIONS.TYPE.REPORTPREVIEW) {
                return false;
            }
            const iouReportID = ReportActionsUtils.getIOUReportIDFromReportActionPreview(action);
            const iouReport = ReportUtils.getReport(iouReportID);
            const shouldShowPayButton = IOU.canIOUBePaid(iouReport, report, policy);
            const shouldShowApproveButton = IOU.canApproveIOU(iouReport, report, policy);
            return shouldShowPayButton || shouldShowApproveButton
        });
        if(shouldCallOpenWalletPage && !didCallOpenWalletPage.current){
            PaymentMethods.openWalletPage();
        }
    }, [reportActions, report]);
```
### What alternative solutions did you explore? (Optional)

- NA

**Comment by bernhardoj** [comment]
This is caused by https://github.com/Expensify/App/pull/30269 where the chat FlatList is remounted multiple times because the listID is renewed multiple times.
https://github.com/Expensify/App/blob/9c8ac2975aa898a2fe90da287a4e11b187b4e5b1/src/pages/home/report/ReportActionsView.tsx#L129-L135

**Comment by youssef-lr** [comment]
cc @perunt @ishpaul777

**Comment by nkdengineer** [comment]
This issue appear in prod as well

**Comment by youssef-lr** [comment]
Here's staging @nkdengineer, do you have a recording for prod?

https://github.com/Expensify/App/assets/9680864/9eb28e18-ea66-4108-8501-287ba5c3c0b0

**Comment by nkdengineer** [comment]
@youssef-lr here is the video in prod (https://new.expensify.com/):

https://github.com/Expensify/App/assets/161821005/1aa2e86e-9a81-4736-abcf-d8a918c2ae80

**Comment by youssef-lr** [comment]
Can you try staging please?

**Comment by nkdengineer** [comment]
@youssef-lr Here is staging:

https://github.com/Expensify/App/assets/161821005/3e8308b2-a451-4d8a-bedd-7f26753b1afa

**Comment by bernhardoj** [comment]
I think it's not deployed to staging yet. On dev, even if you have one button only, it will trigger multiple OpenPaymentsPage calls.

On staging/prod, it will trigger once for each button.

**Comment by sakluger** [comment]
@bernhardoj if this is related to https://github.com/Expensify/App/pull/30269 as you suspect, that PR has a bunch of comments around potential regressions it has caused. I've asked [here in Slack](https://expensify.slack.com/archives/C049HHMV9SM/p1711384398109909?thread_ts=1711147499.521539&cid=C049HHMV9SM) if we should revert it, and I'm going to put this issue on hold while we discuss.\

**Comment by sakluger** [comment]
@roryabraham @perunt @ishpaul777 could you take a look at this one please?

**Comment by sakluger** [comment]
Bump @roryabraham @perunt @ishpaul777

**Comment by perunt** [comment]
There were some fixes implemented related to re-rendering the chat, which should help in this situation as well.
Also, since the bug was captured two weeks ago on production, I can carefully assume that you're receiving so many requests because there are a lot of MoneyRequest instances on the page? I mean, are we sure that these were batched before? It's definitely a bad way to handle it right now, but I can assume that this is how we are handling it currently, and it's not a bug, but it is definitely the wrong approach

**Comment by sakluger** [comment]
Adding a hold and moving to weekly while we figure out what to do with the issue.

**Comment by sakluger** [comment]
Still on hold. Any thoughts on how we should proceed @roryabraham? Should this be added to a project?

**Comment by sakluger** [comment]
Still on hold I guess.

**Comment by sakluger** [comment]
Asked again in Slack which project we should add this to.

**Comment by sakluger** [comment]
Hey @stephanieelliott, assigning to you while I'm OOO until May 31. I'm not really sure what to do with this issue - I've [posted in Slack](https://expensify.slack.com/archives/C049HHMV9SM/p1711384398109909?thread_ts=1711147499.521539&cid=C049HHMV9SM) a few times asking how to move forward and people agree that we should fix it, but I haven't gotten consensus on how to move it forward. I think we can probably just leave it for now and bump in Slack every few weeks.

**Comment by stephanieelliott** [comment]
Roger that, will do @sakluger!

**Comment by stephanieelliott** [comment]
Still holding on this, we're not quite at the `polish` stage of this wave, we'll prioritize it at that time

**Comment by sakluger** [comment]
Still on hold

**Comment by sakluger** [comment]
Still on hold.

**Comment by sakluger** [comment]
I guess we're still on hold.

**Comment by sakluger** [comment]
I asked again in Slack how we can move the issue forward.

**Comment by roryabraham** [comment]
I just retested this on prod. `OpenPaymentsPage` is only used in the Wallet page. Opening that page, I see `OpenPaymentsPage` being called only once. So I'm going to close this.
